### PR TITLE
bpo-31045: Language switch

### PR DESCRIPTION
**Pull request by JulienPalard**
This is the implementation of [this step](https://www.python.org/dev/peps/pep-0545/#create-sphinx-doc-language-switcher) from the PEP 545: make a language switcher in the doc, it looks like this:

documentation language switcher

I tried hard to make this robust, as there's now many cases, and readable, but my JS-foo is old...

I used `Object.keys` which is not compatible before IE 9, but according to a grep sphinx uses underscore.js which seems to use it too so it looked OK but tell me if it is not.

https://bugs.python.org/issue31045

### CODE REVIEW

**Comment by JulienPalard** [comment]
Please don't merge it while we're missing some redirections like: https://docs.python.org/fr/ to https://docs.python.org/fr/3/, directory listings are ugly.

File: Doc/tools/static/switchers.js
Comment (vstinner) [inline Doc/tools/static/switchers.js:94]: most of the code is copied from on_version_switch(), can't you try to factorize the code a little bit more?

File: Doc/tools/static/switchers.js
Comment (vstinner) [inline Doc/tools/static/switchers.js]: can you put docs.python.org in a global constant? It may be useful to have a different value for the doc rendered for readthedocs.org, no? In this case, the ".org/" hack doesn't work anymore :-) Readthedocs.org uses readthedocs.io (not .org) for user docs.

File: Doc/tools/static/switchers.js
Comment (vstinner) [inline Doc/tools/static/switchers.js]: I don't understand the .replace() here. What does it do? Why is it needed?

File: Doc/tools/static/switchers.js
Comment (vstinner) [inline Doc/tools/static/switchers.js]: I would prefer to not hardcode versions here neither. Please put py3k, dev, etc. in a global constant array at the top. By the way, I don't think that "py3k" is still needed.

You can please document in a comment which kinds of versions are accepted here? Example:

// Version examples: "/3", "/dev", "/release/2.7" or "/3.6rc2"

I'm not sure about "/3.6rc2": I wanted to test https://docs.python.org/3.6rc2/ but the HTTP server is sick and fails with HTTP 503 Service Unavailable :-(

**Review by vstinner — CHANGES_REQUESTED** [review]

File: Doc/tools/static/switchers.js
Comment (JulienPalard) [reply Doc/tools/static/switchers.js:94]: Yes I could factorize it, I actually though about it a few seconds, but I don't see how it can make code any better, in the "more readable" sense. Yes it can probably be made shorter (not even sure). But I like the readability of:
```
var new_url = url.replace('.org/' + current_language + current_version,
                          '.org/' + selected_language + current_version);
```
and
```
var new_url = url.replace('.org/' + current_language + current_version,
                          '.org/' + current_language + selected_version);
```

and factorizing it will put a new level of "indirection", making it ultimately harder to read.

File: Doc/tools/static/switchers.js
Comment (JulienPalard) [reply Doc/tools/static/switchers.js]: I pushed an enhancement about the readability of the regex, but I'll have to try later the existance/unexistance of urls like /py3k/ and so on. According to a google search https://docs.python.org/dev/py3k may even exist.

File: Doc/tools/static/switchers.js
Comment (JulienPalard) [reply Doc/tools/static/switchers.js]: I disambiguated it by renaming functions and creating an intermediate named variable.

`find_language_in_url`, now called `language_segment_from_url` extract a path segment like `3.6/`, not a language tag, so we have to strip the `/` to get the language tag.

This is usefull to disambiguate between having no language in URL being the empty string versus having a language in the URL being "LANGUAGE_TAG/", so we can later concatenate "language + version" and it gives the expected result:

- '' + "3.5/" → "3.5/"
- "fr/" + "" → "fr/"
- "fr/" + "3.5/" → "fr/3.5/"

this permit to easily find and replace an optional language and an optional version in an URL like:
```
var new_url = url.replace('.org/' + current_language + current_version,
                          '.org/' + selected_language + current_version);
```

**Comment by vstinner** [comment]
About "https://docs.python.org/py3k": IMHO this URL is redirected to https://docs.python.org/3/" before your Javascript is called, maybe using an Apache Rewrite rule or something like that.

**Comment by JulienPalard** [comment]
Yes I'll have to review all symlinks / redirections before merging this. I also think this can be enhanced like /3/ should probably be a redirection, not a symlink.

**Comment by JulienPalard** [comment]
FTR: I'm waiting for https://github.com/python/psf-salt/pull/107

**Comment by vstinner** [comment]
> FTR: I'm waiting for python/psf-salt#107

This PR has been merged.

**Comment by JulienPalard** [comment]
Yes, I also removed the useless py3k from the regex as it's redirected server side, I'm now setting a full test environment locally to recheck everything, then I'll remove the WIP flag.

**Comment by vstinner** [comment]
Please open an issue at http://bugs.python.org/ and add its number in your PR title.

**Comment by JulienPalard** [comment]
I enhanced the fallback mechanism on unexisting translations, typically switching from `fr/3.7/tutorial.html` to `3.3` now keeps the `tutorial.html` part while removing the `/fr/` as `3.3` is not translated in french, which is better than landing on `3.3/` (loosing the path).

I'll keep the [WIP] tag until [3.7 is built for french on production servers](https://github.com/python/docsbuild-scripts/pull/17).

**Comment by JulienPalard** [comment]
3.7 now also successfully build on docs.python.org/fr/3.7/, I'm removing the [WIP] tag, from my point of view this can be merged.

**Review by vstinner — APPROVED** [review]

**Comment by vstinner** [comment]
@JulienPalard now has access to server logs and so should be able to debug if something goes wrong. The PEP 545 has been approved and the change LGTM. Well, the code is not perfect, but it's not worse than before :-) It can be enhanced later if needed.

**Pull request merged**
